### PR TITLE
feat: bitmap

### LINKED ISSUE #77: [Feature]: BitMaps
### What is the feature you would like to see?

BitMaps library contract ([solidity](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/3e6c86392c97fbc30d3d20a378a6f58beba08eba/contracts/utils/structs/BitMaps.sol))

### PR DESCRIPTION
**Pull request by qalisander**
Resolves #77

- [x] Tests
- [x] Documentation

### CODE REVIEW

**Review by alexfertel — COMMENTED** [review]
Looks good at a first glance, will do a proper review when I get some time.

One question I have is that I think this is a good candidate for implementing `StorageType`, right? What do you think?

File: contracts/src/utils/structs/bitmap.rs
Comment (bidzyyys) [inline contracts/src/utils/structs/bitmap.rs:1]: `U256` instead of `uint256`

File: contracts/src/utils/structs/bitmap.rs
Comment (bidzyyys) [inline contracts/src/utils/structs/bitmap.rs:5]: `U256`

File: contracts/src/utils/structs/bitmap.rs
Comment (bidzyyys) [inline contracts/src/utils/structs/bitmap.rs:6]: Here you have `'bool'`, in other places you have just `bool`

File: contracts/src/utils/structs/bitmap.rs
Comment (bidzyyys) [inline contracts/src/utils/structs/bitmap.rs:57]: Most of the code for `set` and `unset` is common, only last line differs. IMHO you can reuse the code.

**Review by bidzyyys — COMMENTED** [review]
Looks good ✅

**Comment by qalisander** [comment]
> One question I have is that I think this is a good candidate for implementing `StorageType`, right? What do you think?
I think yep. I can reuse this contract to build consecutive on top. We don't need to use stylus inheritance just aggregation

File: contracts/src/utils/structs/bitmap.rs
Comment (qalisander) [reply contracts/src/utils/structs/bitmap.rs]: left just "boolean" word

File: contracts/src/utils/structs/bitmap.rs
Comment (qalisander) [reply contracts/src/utils/structs/bitmap.rs:57]: Agree. It was exactly like that at solidity
Added methods `get_mask` and `get_bucket` to reuse a bit of code

File: contracts/src/utils/structs/bitmap.rs
Comment (qalisander) [inline contracts/src/utils/structs/bitmap.rs:112]: This one seems a bit debatable decision, but now `#[motsu::test]` macro locks storage for a test execution even if there is no contract argument.
Assuming that rust contract struct can be created during test.
And the state of this contract will be single per test.
What do you think @alexfertel ? Or may be use `with_context` function?

File: Cargo.toml
Comment (bidzyyys) [inline Cargo.toml:1]: @alexfertel @qalisander we need to have consistent formatting for `toml` files.

File: contracts/Cargo.toml
Comment (bidzyyys) [inline contracts/Cargo.toml:20]: I believe alloy is configured in main `Cargo.toml`

File: contracts/src/utils/structs/bitmap.rs
Comment (bidzyyys) [inline contracts/src/utils/structs/bitmap.rs:102]: You should use our new `DefaultStorageLayout`, check this: https://github.com/OpenZeppelin/rust-contracts-stylus/commit/9b1c54d96fc865eca6c6934630d636803e1bb49a

File: contracts/src/utils/structs/bitmap.rs
Comment (bidzyyys) [inline contracts/src/utils/structs/bitmap.rs:33]: imho you can reuse here `self.get_mask(index)`

File: contracts/src/utils/structs/bitmap.rs
Comment (bidzyyys) [inline contracts/src/utils/structs/bitmap.rs:32]: you can reuse here `self.get_bucket(index)`

File: contracts/src/utils/structs/bitmap.rs
Comment (bidzyyys) [inline contracts/src/utils/structs/bitmap.rs:58]:
```rust
let bucket = Self::get_bucket(index);
let mask = Self::get_mask(index);
let mut value = self._data.setter(bucket);
let prev = value.get();
```
Is common code for `set` and `unset`

File: contracts/src/utils/structs/bitmap.rs
Comment (qalisander) [reply contracts/src/utils/structs/bitmap.rs:33]: thnx!

File: contracts/src/utils/structs/bitmap.rs
Comment (qalisander) [reply contracts/src/utils/structs/bitmap.rs:102]: nice that it was merged!

File: contracts/Cargo.toml
Comment (qalisander) [reply contracts/Cargo.toml:20]: It was used without arbitrary feature. Since we're using resolver = 2 we can plugin features necessary just for testing purposes

File: contracts/src/utils/structs/bitmap.rs
Comment (qalisander) [reply contracts/src/utils/structs/bitmap.rs:59]: I would say yep, but I don't think we will win much code lanes refactoring it further. It is still looks like solidity counterpart

Comment (qalisander) [reply contracts/src/utils/structs/bitmap.rs:59]: https://github.com/OpenZeppelin/openzeppelin-contracts/blob/53b5d84212b73388a2a7096ff503693dae692726/contracts/utils/structs/BitMaps.sol#L18

File: Cargo.toml
Comment (qalisander) [reply Cargo.toml:1]: yep. I'll remove auto format for `toml` files in my ide:)

File: contracts/src/utils/structs/bitmap.rs
Comment (alexfertel) [reply contracts/src/utils/structs/bitmap.rs:59]: I think it's fine to keep it duplicated, it's only in two places, and we won't be reusing that code anywhere else.

File: .gitignore
Comment (alexfertel) [inline .gitignore:13]: We shouldn't ignore these! It makes failure cases load faster the next time tests run 👀

File: contracts/src/utils/structs/bitmap.rs
Comment (alexfertel) [inline contracts/src/utils/structs/bitmap.rs:12]:
```suggestion
//! Contract module for managing `U256` to boolean mapping in a compact and
//! efficient way, provided the keys are sequential. Largely inspired by Uniswap's 
//! [merkle-distributor].
//!
//! `BitMap` packs 256 booleans across each bit of a single 256-bit slot of
//! `U256` type. Hence, booleans corresponding to 256 _sequential_ indices
//! would only consume a single slot, unlike the regular boolean which would
//! consume an entire slot for a single value.
//!
//! This results in gas savings in two ways:
//!
//! - Setting a zero value to non-zero only once every 256 times
//! - Accessing the same warm slot for every 256 _sequential_ indices
//!
//! [merkle-distributor]: https://github.com/Uniswap/merkle-distributor/blob/master/contracts/MerkleDistributor.sol
```

File: contracts/src/utils/structs/bitmap.rs
Comment (alexfertel) [inline contracts/src/utils/structs/bitmap.rs:30]:
```suggestion
    /// * `index` - index of the boolean value in the bit map.
```

File: contracts/src/utils/structs/bitmap.rs
Comment (alexfertel) [inline contracts/src/utils/structs/bitmap.rs:44]:
```suggestion
    /// * `index` - index of boolean value in the bit map.
    /// * `value` - boolean value to set in the bit map.
```

File: contracts/src/utils/structs/bitmap.rs
Comment (alexfertel) [inline contracts/src/utils/structs/bitmap.rs:81]: Maybe extract `U256::from(1)` and `U256::from(0xff)` as constants to make this a bit faster.

```rust
pub const ONE: U256 = U256::from_limbs([1, 0, 0, 0]);
pub const UNSURE_WHAT_NAME: U256 = U256::from_limbs([255, 0, 0, 0]);
```

File: contracts/src/utils/structs/bitmap.rs
Comment (alexfertel) [inline contracts/src/utils/structs/bitmap.rs:85]: Should we `#[inline]` these? They are private & small.

File: contracts/src/utils/structs/bitmap.rs
Comment (alexfertel) [reply contracts/src/utils/structs/bitmap.rs:106]: I think `with_context` is slightly cleaner because we are instantiating the contract twice this way (with `default`), right? 

In any case, I'm fine either way, I'll leave it to your judgment!

File: contracts/src/lib.rs
Comment (alexfertel) [inline contracts/src/lib.rs:7]: Are you sure we need `core`? It's my understanding that this is always imported.

**Review by alexfertel — APPROVED** [review]
Looks great! 🚀 

Besides the question about `StorageType` I left a few small changes. In any case, I think it's fine to merge as is, and later update it if we need to.

File: contracts/src/lib.rs
Comment (qalisander) [reply contracts/src/lib.rs:7]: weird how it appeared..

File: .gitignore
Comment (qalisander) [reply .gitignore:13]: make sense

File: contracts/src/utils/structs/bitmap.rs
Comment (qalisander) [reply contracts/src/utils/structs/bitmap.rs:88]: I used to think that those small and private functions should be unlined by default

**Pull request merged**
